admin/services/pupil-upload.service.js

```diff
@@ -0,0 +1,49 @@
+const csv = require('fast-csv')
+const fs = require('fs-extra')
+const pupilDataService = require('./data-access/pupil.data.service')
+const validateCSVService = require('./validate-csv.service')
+const generateErrorCSVService = require('./generate-error-csv.service')
+
+const onCSVReadComplete = async (csvDataArray, school) => {
+  const { pupils, csvData, headers, validationErrors, hasValidationError } = await validateCSVService.process(csvDataArray, school)
+  if (hasValidationError) return ({ fileErrors: validationErrors, hasValidationError })
+  // Generate csv with errors
+  const csvHasErrors = csvData.some(p => p[6])
+  if (csvHasErrors) {
+    const csvBlob = await generateErrorCSVService.generate(school, headers, csvData)
+    if (csvBlob.hasError) return (csvBlob.error)
+    return ({
+      csvErrorFile: csvBlob.file.name,
+      hasValidationError: true
+    })
+  } else {
+    // Save pupils if validation is successful
+    let pupilIds = []
+    let savedPupils
+    try {
+      savedPupils = await pupilDataService.insertMany(pupils)
+    } catch (error) {
+      return (error)
```

Comment:
How does the caller differentiate between an error and the actual result?

Reply:
the caller should catch the exception coming from mongoose so that try/catch block is not needed.
I am removing it.

Reply:
Since the caller (pupil controller) is listening for errors I suppose that try/catch is redundant.
The only case that is not captured is when no pupils are saved in which case a separate error must be thrown.